crypt.go

```diff
@@ -0,0 +1,488 @@
+// field of options contains the "tries=" option, then an error will be returned. This option cannot be used with this function.
+//
+// If the LockSealedKeyAccess field of options is true and the call to LockAccessToSealedKeys fails, a LockAccessToSealedKeysError
+// error will be returned. In this case, activation with either the TPM sealed key or the fallback recovery key will not be attempted.
+//
+// If activation with the TPM sealed key fails, a *ActivateWithTPMSealedKeyError error will be returned, even if the subsequent
+// fallback recovery activation is successful. In this case, the RecoveryErr field of the returned error will be nil, and the TPMErr
+// field will contain the original error. If activation with the fallback recovery key also fails, the RecoveryErr field of the
+// returned error will also contain details of the error encountered during recovery key activation.
+//
+// If the volume is successfully activated, either with the TPM sealed key or the fallback recovery key, this function returns true.
+// If it is not successfully activated, then this function returns false.
+func ActivateVolumeWithTPMSealedKey(tpm *TPMConnection, volumeName, sourceDevicePath, keyPath string, pinReader io.Reader, options *ActivateWithTPMSealedKeyOptions) (bool, error) {
+	if options.PINTries < 0 {
+		return false, errors.New("invalid PINTries")
+	}
+	if options.RecoveryKeyTries < 0 {
+		return false, errors.New("invalid RecoveryKeyTries")
+	}
+
+	activateOptions, err := makeActivateOptions(options.ActivateOptions)
+	if err != nil {
+		return false, err
+	}
+
+	if err := activateWithTPMKey(tpm, volumeName, sourceDevicePath, keyPath, pinReader, options.PINTries, options.LockSealedKeyAccess, activateOptions); err != nil {
+		reason := RecoveryReasonUnexpectedError
+		switch {
+		case isLockAccessError(err):
+			return false, LockAccessToSealedKeysError(err.Error())
+		case xerrors.Is(err, ErrTPMLockout):
+			reason = RecoveryReasonTPMLockout
+		case xerrors.Is(err, ErrTPMProvisioning):
+			reason = RecoveryReasonTPMProvisioningError
+		case isInvalidKeyFileError(err):
+			reason = RecoveryReasonInvalidKeyFile
+		case xerrors.Is(err, requiresPinErr):
+			reason = RecoveryReasonPINFail
+		case xerrors.Is(err, ErrPINFail):
+			reason = RecoveryReasonPINFail
+		case isExecError(err, systemdCryptsetupPath):
+			// systemd-cryptsetup only provides 2 exit codes - success or fail - so we don't know the reason it failed yet. If activation
+			// with the recovery key is successful, then it's safe to assume that it failed because the key unsealed from the TPM is incorrect.
+			reason = RecoveryReasonInvalidKeyFile
+		}
+		rErr := activateWithRecoveryKey(volumeName, sourceDevicePath, nil, options.RecoveryKeyTries, reason, activateOptions)
+		return rErr == nil, &ActivateWithTPMSealedKeyError{err, rErr}
+	}
+
+	return true, nil
+}
+
+type ActivateWithRecoveryKeyOptions struct {
```

Comment:
this needs doc comments as well.

Reply:
I've added some documentation for this now.